File: c++/HardDrive.cpp
#include <iostream>
using namespace std;
#include "Komponen.cpp"
#include "Compatible.h"

class Harddrive : public Komponen, public Compatible {
private:
    int kapasitasGB;
    string tipeStorage;
    int readSpeedMBps;
    int writeSpeedMBps;

public:
    Harddrive(string merk, int tahunProduksi, double harga, string kode, string tipe, int kapasitasGB, string tipeStorage, int readSpeedMBps, int writeSpeedMBps);
    int getKapasitasGB();
    void setKapasitasGB(int kapasitasGB);
    string getTipeStorage();
    void setTipeStorage(string tipeStorage);
    int getReadSpeedMBps();
    void setReadSpeedMBps(int readSpeedMBps);
    int getWriteSpeedMBps();
    void setWriteSpeedMBps(int writeSpeedMBps);
    double calculatePerformance();
    void displayInfo() override;
    bool checkCompatibility(Komponen* komponen) override;
    string* getCompatibleWith() override;
};

int Harddrive::getKapasitasGB() { return kapasitasGB; }
void Harddrive::setKapasitasGB(int kapasitasGB) { this->kapasitasGB = kapasitasGB; }

string Harddrive::getTipeStorage() { return tipeStorage; }
void Harddrive::setTipeStorage(string tipeStorage) { this->tipeStorage = tipeStorage; }

int Harddrive::getReadSpeedMBps() { return readSpeedMBps; }
void Harddrive::setReadSpeedMBps(int readSpeedMBps) { this->readSpeedMBps = readSpeedMBps; }

int Harddrive::getWriteSpeedMBps() { return writeSpeedMBps; }
void Harddrive::setWriteSpeedMBps(int writeSpeedMBps) { this->writeSpeedMBps = writeSpeedMBps; }

double Harddrive::calculatePerformance() {
    return (readSpeedMBps + writeSpeedMBps) / 2.0;
}

void Harddrive::displayInfo() {
    Komponen::displayInfo();
    cout << "Kapasitas GB: " << kapasitasGB << endl;
    cout << "Tipe Storage: " << tipeStorage << endl;
    cout << "Read Speed MBps: " << readSpeedMBps << endl;
    cout << "Write Speed MBps: " << writeSpeedMBps << endl;
}

bool Harddrive::checkCompatibility(Komponen* komponen) {
    // Dummy implementation
    return true;
}

string* Harddrive::getCompatibleWith() {
    // Dummy implementation
    static string compatibleWith[] = {"SATA", "NVMe"};
    return compatibleWith;
}

File: java/Harddrive.java
public class Harddrive extends Komponen implements Compatible {
    private int kapasitasGB;
    private String tipeStorage;
    private int readSpeedMBps;
    private int writeSpeedMBps;

    public Harddrive(String merk, int tahunProduksi, double harga, String kode, String tipe, int kapasitasGB, String tipeStorage, int readSpeedMBps, int writeSpeedMBps) {
        super(merk, tahunProduksi, harga, kode, tipe);
        this.kapasitasGB = kapasitasGB;
        this.tipeStorage = tipeStorage;
        this.readSpeedMBps = readSpeedMBps;
        this.writeSpeedMBps = writeSpeedMBps;
    }

    public int getKapasitasGB() { return kapasitasGB; }
    public void setKapasitasGB(int kapasitasGB) { this.kapasitasGB = kapasitasGB; }

    public String getTipeStorage() { return tipeStorage; }
    public void setTipeStorage(String tipeStorage) { this.tipeStorage = tipeStorage; }

    public int getReadSpeedMBps() { return readSpeedMBps; }
    public void setReadSpeedMBps(int readSpeedMBps) { this.readSpeedMBps = readSpeedMBps; }

    public int getWriteSpeedMBps() { return writeSpeedMBps; }
    public void setWriteSpeedMBps(int writeSpeedMBps) { this.writeSpeedMBps = writeSpeedMBps; }

    public double calculatePerformance() {
        return (readSpeedMBps + writeSpeedMBps) / 2.0;
    }

    @Override
    public void displayInfo() {
        super.displayInfo();
        System.out.println("Kapasitas GB: " + kapasitasGB);
        System.out.println("Tipe Storage: " + tipeStorage);
        System.out.println("Read Speed MBps: " + readSpeedMBps);
        System.out.println("Write Speed MBps: " + writeSpeedMBps);
    }

    
    public boolean checkCompatibility(Komponen komponen) {
        // Dummy implementation
        return true;
    }

    
    public String[] getCompatibleWith() {
        // Dummy implementation
        return new String[]{"SATA", "NVMe"};
    }
}